plugins/chat/app/services/chat/add_users_to_channel.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible to add users to a channel.
+  # The guardian passed in is the "acting user" when adding users.
+  # The service is essentially creating memberships for the users.
+  #
+  # @example
+  #  ::Chat::AddUsersToChannel.call(
+  #    guardian: guardian,
+  #    channel_id: 1,
+  #    usernames: ["bob", "alice"]
+  #  )
+  #
+  class AddUsersToChannel
+    include Service::Base
+
+    # @!method call(guardian:, **params_to_create)
+    #   @param [Guardian] guardian
+    #   @param [Integer] id of the channel
+    #   @param [Hash] params_to_create
+    #   @option params_to_create [Array<String>] usernames
+    #   @return [Service::Base::Context]
+    contract
+    model :channel
+    policy :can_add_users_to_channel
+    model :users
+
+    transaction do
+      step :upsert_memberships
+      step :recompute_users_count
+      step :notice_channel
+    end
+
+    # @!visibility private
+    class Contract
+      attribute :usernames, :array
+      validates :usernames, presence: true
+
+      attribute :channel_id, :integer
+      validates :channel_id, presence: true
+
+      validate :usernames_length
+
+      def usernames_length
+        if usernames && usernames.length > SiteSetting.chat_max_direct_message_users + 1 # 1 for current user
+          errors.add(
+            :usernames,
+            "should have less than #{SiteSetting.chat_max_direct_message_users} elements",
+          )
+        end
+      end
+    end
+
+    private
+
+    def can_add_users_to_channel(guardian:, channel:, **)
+      (guardian.user.admin? || channel.joined_by?(guardian.user)) &&
+        channel.direct_message_channel? && channel.chatable.group
```

Comment:
Not that for now this service can only be used with direct message channels, but I decided to not show it in the name as I think it will change soon, could still reconsider